osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java

```diff
@@ -69,23 +70,48 @@ public InstallationService() {
   public void addMeter(
       final MessageMetadata messageMetadata, final AddSmartMeterRequest addSmartMeterRequest)
       throws FunctionalException {
+    this.storeMeter(messageMetadata, addSmartMeterRequest);
+    this.sendRequestToOsgpCore(messageMetadata, addSmartMeterRequest);
+  }
+
+  private void storeMeter(
+      final MessageMetadata messageMetadata, final AddSmartMeterRequest addSmartMeterRequest)
+      throws FunctionalException {
     final String organisationId = messageMetadata.getOrganisationIdentification();
     final String deviceId = messageMetadata.getDeviceIdentification();
     LOGGER.debug(
-        "addMeter for organisationIdentification: {} for deviceIdentification: {}",
+        "store meter for organisationIdentification: {} for deviceIdentification: {}",
         organisationId,
         deviceId);
     final SmartMeteringDevice smartMeteringDevice = addSmartMeterRequest.getDevice();
 
     final String deviceIdentification = messageMetadata.getDeviceIdentification();
-
-    this.smartMeterService.validateSmartMeterDoesNotExist(deviceIdentification);
+    final Optional<SmartMeter> existingSmartMeter =
+        this.smartMeterService.validateSmartMeterDoesNotExist(
+            deviceIdentification, addSmartMeterRequest.getOverwrite());
 
     final SmartMeter smartMeter = this.smartMeterService.convertSmartMeter(smartMeteringDevice);
+    if (existingSmartMeter.isPresent()) {
+      if (addSmartMeterRequest.getOverwrite()) {
+        smartMeter.setVersion(existingSmartMeter.get().getVersion());
+        smartMeter.setId(existingSmartMeter.get().getId());
+      } else {
+        LOGGER.error(
+            "ERROR: SmartMeter with device identification {} already exists and overwrite is not enabled",
+            messageMetadata.getDeviceIdentification());
+      }
```

Comment:
This is already covered by the _validateSmartMeterDoesNotExist_ method, right?

Reply:
Right, removed duplicate code and updated method name to checkIfSmartMeterExistsAndOverwriteAllowed